aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+which store a local vector to preserve the lifetime of data that's viewed by
+TensorLists further down the call chain.
+*******************************************************************************/
+
+// helper to pick out the TensorList arg and create a casted vector
+template<typename Arg0, typename... Args> inline std::vector<Tensor>
+get_casted_vector(at::ScalarType to_type, Arg0 arg0, Args... args) {
+  return get_casted_vector(args...);
+}
+
+// Specialization for the TensorList itself
+template<typename... Args> inline std::vector<Tensor>
+get_casted_vector(at::ScalarType to_type, const TensorList& list, Args... args) {
+  return cached_cast(to_type, list);
+}
+
+// Overload to catch Tensor args
+inline Tensor
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, const Tensor& arg) {
+  return cached_cast(to_type, arg);
+}
+
+// Overload to catch the TensorList arg and swap in our vector
+inline const std::vector<Tensor>&
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, const TensorList& arg) {
+  return casted_vector;
+}
+
+// Template to catch other args.
+template<typename T> inline T
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, T arg) {
+  return arg;
+}
+
+/*******************************************************
+Logic to flip an output dtype flag.
+Keep it simple for now by assuming only one such flag is
+present in the argument list.  If I ever need a function
+with more than flag I'll figure out something else.
+The policy is:
+If the user has explicity specified a dtype, respect it.
+Otherwise, set it to the autocast type.
+********************************************************/
+
+// Overload to catch dtype flags
+inline c10::optional<ScalarType> set_opt_dtype(at::ScalarType to_type, const c10::optional<ScalarType>& dtype) {
+  return dtype.has_value() ? dtype : to_type;
+}
+
+// Template to catch other args
+template<typename T>
+inline T set_opt_dtype(at::ScalarType to_type, T arg) {
+  return arg;
+}
+
+/********************************************************************************************************
+Templates to provide wrapper functions
+
+I'm copying the pattern used in core/boxing/kernel_function.h to extract args and return type.
+(see also https://stackoverflow.com/questions/46533698/how-to-deduce-argument-list-from-function-pointer)
+
+This strategy uses an exterior "WrapFunction" that extracts arguments on behalf of
+(in my case several specializations of) an interior "WrapFunction_".
+Interior WrapFunction_ specializations are defined for each CastPolicy.
+********************************************************************************************************/
+
+// Base template for WrapFunction_, which is specialized to host call methods for each CastPolicy
+template<CastPolicy policy, class Redispatch, Redispatch* F, class Ret, class ArgList> struct WrapFunction_ {};
```

Comment:
I am wondering why `Ret` and `ArgList` can't be inferred from `Redispatch`. If the reason is because the top-level function doesn't exactly match the native function, it seems to me that we should offer some sort of uniform interface which does match. (Not actionable for this PR, but something to think about.) @smessmer also take a look at this please!

Reply:
Hmm maybe I can, sometimes the necessary-and-sufficient set of template parameters is hard to tell.  I'll think about it...